### PR TITLE
test(NODE-3188): backport transaction pinning tests

### PR DESCRIPTION
**Pull request by durran**
Backport of https://github.com/mongodb/node-mongodb-native/pull/2831

### CODE REVIEW

**Review by emadum — APPROVED** [review]
There are some unrelated tests failing in CI, could you skip them and reference NODE-3308 so we can have a green build? Otherwise LGTM!

File: test/manual/atlas_connectivity.test.js [hunk 1]
```diff
@@ -15,8 +15,15 @@ describe('Atlas Connectivity', function() {
     context(configName, function() {
       CONFIGS[configName].forEach(connectionString => {
         const name = connectionString.indexOf('mongodb+srv') >= 0 ? 'mongodb+srv' : 'normal';
-        it(`${name} (unified)`, makeConnectionTest(connectionString, { useUnifiedTopology: true }));
-        it(`${name} (legacy)`, makeConnectionTest(connectionString, { useUnifiedTopology: false }));
+        // TODO: Skipping until NODE-3308 is merged.
```
Comment (emadum) [inline test/manual/atlas_connectivity.test.js:18]: Good call skipping these too, but NODE-3308 is actually referring to different tests: the `A successful find event with a getmore and the server kills the cursor` failures.

**Review by emadum — CHANGES_REQUESTED** [review]

File: lib/cursor.js [hunk 2]
```diff
@@ -165,6 +165,10 @@ class Cursor extends CoreCursor {
     return this.cmd.sort;
   }
 
+  set session(clientSession) {
```
Comment (nbbeeken) [inline lib/cursor.js:168]: Could you bring in the helper from 4.0 and use that so we can keep setting this private?

Here it is for convenience: (I put it in `test/tools/utils.js`)
```javascript
function getSymbolFrom(target, symbolName, assertExists = true) {
  const symbol = Object.getOwnPropertySymbols(target).filter(
    s => s.toString() === `Symbol(${symbolName})`
  )[0];

  if (assertExists && !symbol) {
    throw new Error(`Did not find Symbol(${symbolName}) on ${target}`);
  }

  return symbol;
}
```

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: lib/cursor.js (hunk 2 above)
Comment (durran) [reply lib/cursor.js:168]: Per our discussion I just refactored the same as I did on the 4.0 branch.

File: test/functional/unified-spec-runner/operations.ts [hunk 3]
```diff
@@ -79,6 +79,18 @@ interface OperationFunctionParams {
 type RunOperationFn = (p: OperationFunctionParams) => Promise<Document | boolean | number | void>;
 export const operations = new Map<string, RunOperationFn>();
 
+function executeWithPotentialSession(
+  entities: EntitiesMap,
+  operation: OperationDescription,
+  cursor: AbstractCursor
```
[the remark this answers is not in the document]
Comment (durran) [reply test/functional/unified-spec-runner/operations.ts]: Used `CoreCursor` here so ts could recognise the `toArray` method now.

File: test/manual/atlas_connectivity.test.js (hunk 1 above)
Comment (durran) [reply test/manual/atlas_connectivity.test.js:32]: All tests passing now.

**Review by emadum — APPROVED** [review]

**Review by nbbeeken — APPROVED** [review]

**Review by dariakp — APPROVED** [review]

**Pull request merged**